src/bootstrap/bin/main.rs

```diff
@@ -7,12 +7,28 @@
 
 use std::env;
 
-use bootstrap::{Build, Config, Subcommand, VERSION};
+use bootstrap::{t, Build, Config, Subcommand, VERSION};
 
 fn main() {
     let args = env::args().skip(1).collect::<Vec<_>>();
     let config = Config::parse(&args);
 
+    let mut build_lock;
+    let _build_lock_guard;
+    if cfg!(any(unix, windows)) {
+        build_lock = fd_lock::RwLock::new(t!(std::fs::File::create(config.out.join("lock"))));
+        _build_lock_guard = match build_lock.try_write() {
+            Ok(lock) => lock,
+            err => {
+                println!("warning: build directory locked, waiting for lock");
+                drop(err);
+                t!(build_lock.write())
```

Comment:
Does the lock guard here get correctly cleared after rustbuild is killed, including with process::exit(...) or ctrl+c?

Both of those are today expected to happen in typical usage, I think.

Reply:
@Mark-Simulacrum Yes, OS file locking doesn't persist after the file is closed by the process, so if the process exits the lock goes away. This is the major advantage of OS file locking over just using the existence of a lockfile.